platform/linux-generic/odp_crypto_ipsecmb.c

```diff
@@ -0,0 +1,999 @@
+			     ODP_CRYPTO_ALG_ERR_NONE);
+}
+
+static void
+null_crypto_routine(odp_packet_t pkt ODP_UNUSED,
+		    const odp_crypto_packet_op_param_t *param ODP_UNUSED,
+		    odp_crypto_generic_session_t *session ODP_UNUSED)
+{
+	set_crypto_op_result_ok(pkt);
+}
+
+
+static
+void zuc_eea3_cipher_op(odp_packet_t pkt,
+		      const odp_crypto_packet_op_param_t *param,
+		      odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->cipher_range.offset;
+	uint32_t in_len = param->cipher_range.length;
+        uint8_t dst_data[MAX_BUFFER_LENGTH_IN_BYTES];
+
+#if ODP_DEPRECATED_API
+	if (param->cipher_iv_ptr)
+		iv_ptr = param->cipher_iv_ptr;
+	else if (session->p.cipher_iv.data)
+		iv_ptr = session->cipher.iv_data;
+	else {
+		set_crypto_op_result(pkt,
+				     ODP_CRYPTO_ALG_ERR_DATA_SIZE,
+				     ODP_CRYPTO_ALG_ERR_NONE);
+		return;
+	}
+#else
+	iv_ptr = param->cipher_iv_ptr;
+	_ODP_ASSERT(session->p.cipher_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(odp_packet_is_segmented(pkt))) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (session->p.cipher_key.length == 16) {
+		/* ZUC128 EEA3 */
+		IMB_ZUC_EEA3_1_BUFFER(mb_mgr, (const void  **)session->cipher.key_data,
+				      (const void *)iv_ptr,
+				      (const void *)data,
+				      (void *)dst_data,
+                                      in_len);
+	} else if (session->p.cipher_key.length == 32) {
+		/* ZUC256 EEA3 */
+		IMB_ZUC256_EEA3_1_BUFFER(mb_mgr, (const void **)session->cipher.key_data,
+					 (const void *)iv_ptr,
+					 (const void *)data,
+					 (void *)dst_data,
+					 in_len);
+	}
+
+	odp_packet_copy_from_mem(pkt, in_pos, in_len, dst_data);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static int process_zuc_eea3_param(odp_crypto_generic_session_t *session)
+{
+	/* Verify key length is valid */
+	if (16 != session->p.cipher_key.length &&
+	    32 != session->p.cipher_key.length)
+		return -1;
+
+	/* Verify the IV length is correct */
+	if (16 != session->p.cipher_iv_len &&
+	    25 != session->p.cipher_iv_len)
+		return -1;
+
+	memcpy(session->cipher.key_data, session->p.cipher_key.data,
+		session->p.cipher_key.length);
+
+	/* Set function */
+	if (ODP_CRYPTO_OP_ENCODE == session->p.op)
+		session->cipher.func = zuc_eea3_cipher_op;
+	else
+		session->cipher.func = zuc_eea3_cipher_op;
+
+	return 0;
+}
+
+static
+void auth_zuc_eia3_gen(odp_packet_t pkt,
+		       const odp_crypto_packet_op_param_t *param,
+		       odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->auth_range.offset;
+	uint32_t in_len = param->auth_range.length;
+        uint8_t dst_data[MAX_BUFFER_LENGTH_IN_BYTES];
```

Comment:
There is no need to have this long buffer just for the authentication tag (which is what goes in here). I also propose renaming this from dst_data to tag (or auth_tag or something like that) since dst_data may give a wrong impression on what it is for.

Reply:
Updated with `auth_tag[ZUC_DIGEST_LENGTH]` (4)

Reply:
Now there are these two macros:
#define IPSEC_MB_CRYPTO_MAX_DIGEST_LENGTH          16
#define ZUC_DIGEST_LENGTH 4

The only supported digest (I would not call it digest since it is not just a hash, but I digress) length is 4 bytes, so the former macro is not so good and apparently a copy paste leftover. But auth_zuc_eia3_check() still uses it instead of the newly defined macro that is now used in auth_zuc_eia3_gen().

Reply:
Removed `IPSEC_MB_CRYPTO_MAX_DIGEST_LENGTH` and use  `ZUC_DIGEST_LENGTH` instead.